src/NLog/Internal/XmlHelper.cs

```diff
@@ -445,5 +445,77 @@ private static string EnsureDecimalPlace(string text)
         }
 
         private static readonly char[] DecimalScientificExponent = new[] { 'e', 'E' };
+
+        internal static bool ContainsInvalidXmlChars(StringBuilder text, int startPos = 0)
+        {
+            for (int i = startPos; i < text.Length; i++)
+            {
+                // Check if this character is the start of a surrogate pair
+                if (char.IsHighSurrogate(text[i]) && i + 1 < text.Length && char.IsLowSurrogate(text[i + 1]))
+                {
+                    int codePoint = char.ConvertToUtf32(text[i], text[i + 1]);
+                    if (!IsValidXmlChar(codePoint))
+                        return true;
+
+                    i++; // skip the low surrogate
+                }
+                else
+                {
+                    int codePoint = text[i];
+                    if (!IsValidXmlChar(codePoint))
+                        return true;
+                }
+            }
+
+            return false;
+        }
+        private static bool IsValidXmlChar(int codePoint)
```

Comment:
Notice there are also similar methods `XmlHelper.RemoveInvalidXmlChars`  + `XmlConvert.IsXmlChar`. Maybe there can be some code-reuse ?